sdk/core/azure-core/azure/core/pipeline/policies/_authentication.py

```diff
@@ -172,12 +172,16 @@ class AzureKeyCredentialPolicy(SansIOHTTPPolicy):
     def __init__(
         self, credential: "AzureKeyCredential", name: str, **kwargs  # pylint: disable=unused-argument
     ) -> None:
-        super(AzureKeyCredentialPolicy, self).__init__()
-        self._credential = credential
+        super().__init__()
+        if not hasattr(credential, "key"):
+            raise TypeError(
+                "String is not a supported credential input type. Use an instance of AzureKeyCredential."
```

Comment:
My only concern is the error message. We do have custom key credential types. Not a blocker, but if anyone can come up with a better name/how to insert the correct credential type name into the string, it would be great. Not sure if it can be done without changing codegen/all callers though.

Reply:
So I assume the custom key will use the same policy? But if the custom key is connected to its own policy, then there is no problem

Reply:
Played a bit with generic, but you cant'get them from `__init__`
https://stackoverflow.com/questions/57706180/generict-base-class-how-to-get-type-of-t-from-within-instance

```python
def get_actual_type_as_str(obj: Any):
    return str(typing.get_args(obj.__orig_class__)[0].__name__)

class Foo(Generic[CredTypeVar]):
    def __init__(self, credentials: CredTypeVar) -> None:
        type = get_actual_type_as_str(self) # Do not work in init
        print(f"I expect {type}")

   def do_something(self):
       type = get_actual_type_as_str(self) # Works here

f = Foo(AzureKeyCredential())
f = Foo[CustomAzureKeyCredential](CustomAzureKeyCredential())

```

So I don't see a way using typing to do it, which would have been neat.